### PR TITLE
Add already seen and retired banners

### LINKED ISSUE #593: Already seen/Finished subject banners
**Package** lib-classifier

**Feature or Issue Description**
We need to add this UI element. The current PFE version causes a lot of confusion, but we still need to inform volunteers on the state of the subject to not waste their time.

We do have an updated design for this:
- Already seen banner: https://projects.invisionapp.com/d/main#/console/12923997/328618227/preview
- Finished banner: 
https://projects.invisionapp.com/d/main#/console/12923997/328618228/preview

These new banners include a 'why' button that shows a tooltip on hover and/or on click?

Also related to #204

### PR DESCRIPTION
**Pull request by rogerhutchings**
Package:

lib-classifier

Closes #593.

Describe your changes:

Adds the already seen and retired banners.

The `Why am I seeing this?` button will show an explanatory tooltip on hover, and on tap if hover is unavailable. The tooltip is dismissed by moving the mouse away, or tapping the tooltip or button again if hover is unavailable.

The tooltip is themed for light and dark.

### Retired banner

gif1

### Already seen banner

gif2

## How to test in browser

1. Open the React devtools
1. Search for `AlreadySeenBannerContainer` or `RetiredBannerContainer`
1. Select the first child, which will be an instance of `Banner`
1. Toggle the `show` prop

## Notes

- In the LCV, if a banner is visible you can't delete an annotation, due to the delete button placement.

    gif3

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?
- [ ] Is the changelog updated?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?

### CODE REVIEW

**Comment by camallen** [comment]
Related to https://github.com/zooniverse/Panoptes-Front-End/issues/4949#issue-364905014

Can these explanatory tools tips reflect the serialized subject information:
1. if a user has finished all the available data - `user_has_finished_workflow`
2. the workflow is finished - `finished_workflow`

Both points relate to when users classifies all the remaining available data: 
1. in the power user case when there is still lots of data available
2. a project only has a few available subjects left but isn't finished

**Comment by rogerhutchings** [comment]
Blork, I'll take a look at those tests.

The reason for wanting it to work without hover was for touch devices 📱

File: packages/lib-classifier/src/components/Classifier/components/Banners/Banners.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { Stack } from 'grommet'
+import React from 'react'
+import styled from 'styled-components'
```
[the remark this answers is not in the document]
Comment (rogerhutchings) [reply packages/lib-classifier/src/components/Classifier/components/Banners/Banners.js]: nope

**Comment by rogerhutchings** [comment]
This isn't a properly multidimensional table but should be enough to cover the possible states...

|                                   |                       already_seen: true                      |    already_seen: false   |                       retired: true                      |      retired: false      |  finished_workflow: true |                    finished_workflow: false                   |                user_has_finished_workflow: true               | user_has_finished_workflow: false |
|:---------------------------------:|:-------------------------------------------------------------:|:------------------------:|:--------------------------------------------------------:|:------------------------:|:------------------------:|:-------------------------------------------------------------:|:-------------------------------------------------------------:|:---------------------------------:|
|         already_seen: true        |                      Already seen banner                      |    Already seen banner   |                      Retired banner                      |    Already seen banner   | Finished workflow banner |                      Already seen banner                      | Already seen banner (with note saying you've seen everything) |        Already seen banner        |
|        already_seen: false        |                      Already seen banner                      |         No banner        |                      Retired banner                      |         No banner        | Finished workflow banner |                           No banner                           |                               X                               |             No banner             |
|           retired: true           |                         Retired banner                        |      Retired banner      |                      Retired banner                      |             X            | Finished workflow banner |                         Retired banner                        |       Retired (with note saying you've seen everything)       |           Retired banner          |
|           retired: false          |                      Already seen banner                      |         No banner        |                             X                            |         No banner        |             X            |                           No banner                           | Already seen banner (with note saying you've seen everything) |             No banner             |
|      finished_workflow: true      |                    Finished workflow banner                   | Finished workflow banner |                 Finished workflow banner                 | Finished workflow banner | Finished workflow banner |                               X                               |                    Finished workflow banner                   |      Finished workflow banner     |
|      finished_workflow: false     |                      Already seen banner                      |         No banner        |                      Retired banner                      |         No banner        |             X            |                           No banner                           | Already seen banner (with note saying you've seen everything) |             No banner             |
|  user_has_finished_workflow: true | Already seen banner (with note saying you've seen everything) |             X            | Retired banner (with note saying you've seen everything) |    Already seen banner   | Finished workflow banner | Already seen banner (with note saying you've seen everything) | Already seen banner (with note saying you've seen everything) |                 X                 |
| user_has_finished_workflow: false |                      Already seen banner                      |         No banner        |                      Retired banner                      |         No banner        | Finished workflow banner |                           No banner                           |                               X                               |                 X                 |

**Comment by eatyourgreens** [comment]
The hover behaviour is really annoying when moving the mouse around. I'd prefer if 'Why am I seeing this?' was a button that you have to click.

The colour contrast here is too low.
Screenshot showing that 'Why am I seeing this?' text is almost the same colour as the background.

**Comment by rogerhutchings** [comment]
cc @beckyrother ^

1. Happy with changing the 'Why am I seeing this?' colour to white?
1. Hover or click for the tooltip?

**Comment by eatyourgreens** [comment]
VoiceOver announces the button as 'Why A M I seeing this (button)', which is weird. VoiceOver also skips past the popup and never reads it, so I don't actually find out why I'm seeing this.

Unrelated but weird issue while I was going through the page in VoiceOver: VoiceOver announces a form checkbox input somewhere in the project title, in the navigation. I think it's announcing the 'Zooniverse Approved' tick graphic but I'm not sure why it's being parsed as a checkbox input.

**Comment by rogerhutchings** [comment]
@camallen we could simplify this by having four banners:

1. User has completed everything
1. Workflow is finished
1. Subject has been seen
1. Subject has been retired

**Comment by eatyourgreens** [comment]
> I wonder if it's taking an interpretation hint from the capitalisation and thinks it means time?

I'm not sure. It's reading 'Add to collections' as 'A D D to collections' and same with the favourites button.

**Comment by eatyourgreens** [comment]
Tabbing through the page, I can't move past the 'Why am I seeing this?' button. Opening the popup info seems to move focus back to the browser chrome and I find myself tabbing through my open tab controls. That would explain why VoiceOver isn't reading the content of the popup.

**Comment by eatyourgreens** [comment]
I thought a screenshot might help, so here's how the button behaves in VoiceOver. There's no popup text, because it isn't hovered or focussed. VoiceOver does tell me it's a clickable button, and does tell me I've clicked it when I click it.

Screenshot showing the VoiceOver announcement for the 'Why am I seeing this?' button.

**Comment by camallen** [comment]
> @camallen we could simplify this by having four banners:
> 
> 1. User has completed everything
> 2. Workflow is finished
> 3. Subject has been seen
> 4. Subject has been retired

Yep - that works for me 👍

**Comment by rogerhutchings** [comment]
Okay, so the banner precedence is as follows:

1. Workflow finished
1. User has finished workflow
1. Subject retired
1. Subject seen

**Comment by eatyourgreens** [comment]
If `already_seen` is false, do we also check against which subjects you've seen this session? PFE does that but I'm not sure if it's necessary.

**Comment by rogerhutchings** [comment]
As it stands, no. Looking at the `SubjectStore`, I'm not sure it records subjects seen either; it looks like they're just removed from the store entirely.

**Comment by srallen** [comment]
> As it stands, no. Looking at the SubjectStore, I'm not sure it records subjects seen either; it looks like they're just removed from the store entirely.

I think the idea was to push classified subjects into the recents store, so we don't need to maintain two places of 'already seen' subjects.

**Comment by rogerhutchings** [comment]
Ah right. Still, if unless anyone thinks we desperately need it, I'm inclined not to bother; we could also be checking after every subject for whether the workflow is complete, or whether the user has finished the workflow.

**Comment by rogerhutchings** [comment]
@eatyourgreens on accessibility: any ideas why it would be out of regular tab order? It's just a button with its default tabIndex...

**Comment by eatyourgreens** [comment]
It's not out of the regular tab order. I can tab to it just fine. Focus then leaves the window and passes to the browser chrome.

**Comment by rogerhutchings** [comment]
oh wait, it might be the absolute positioning

**Comment by eatyourgreens** [comment]
Is the popup a modal popup? If so, tab would cycle out of the viewport while the modal is open. But the modal closes onBlur and reopens onFocus so you can never tab past it.

The simple answer to this is to remove onHover behaviour and replace it with onClick.

**Comment by rogerhutchings** [comment]
@eatyourgreens okay, try it now

**Comment by eatyourgreens** [comment]
Much better!
Screenshot of the 'Why am I seeing this?' button showing VoiceOver reading the popup help text.

The [Grommet docs for Drop](https://v2.grommet.io/drop) seem to imply that I can close the popup with Esc, but that isn't happening. I guess I toggle it with the spacebar on the button.

**Comment by rogerhutchings** [comment]
Okay, the tooltip now closes on Esc, and is automatically hidden if it's still open when the subject changes and the banner disappears.

File: packages/lib-classifier/src/components/Classifier/components/Banners/components/Banner/Banner.spec.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        })
+      })
+
+      afterEach(function () {
+        wrapper.setState({
+          tooltipOpen: false
+        })
+      })
+
+      it.only('should toggle the tooltip on tap', function () {
```
Comment (srallen) [inline packages/lib-classifier/src/components/Classifier/components/Banners/components/Banner/Banner.spec.js]: Leftover `only`

File: packages/lib-classifier/src/components/Classifier/components/Banners/components/Banner/Banner.js [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+      className,
+      theme: { mode },
+      show,
+      tooltipText
+    } = this.props
+
+    return (
+      <PosedBox
+        align='center'
+        background={transparentize(0.5, background)}
```
[the remark this answers is not in the document]
Comment (rogerhutchings) [reply packages/lib-classifier/src/components/Classifier/components/Banners/components/Banner/Banner.js]: ooh, TIL

**Comment by rogerhutchings** [comment]
Okay, let's change out the opacity then. Side note: @beckyrother, can we do an accessibility review at the design stage?

Animations: I didn't know grommet had animations for `Box`; I used `react-pose` as I wanted to add some other animations to the classifier, but let's see what we can achieve with grommet's animation for now. I chose the the current architecture as it's very easy to test, but I don't mind taking the time to do it another way.

> I noticed while testing that I logged out while one of the banners was showing and the state of the banner wasn't reset... should we refetch for the subject queue when someone logs out or logs in since this state is being serialized by the API and thus all the subjects in the queue may no longer have accurate states when after logging out? (this is a separate issue and can be addressed in another PR)

Yup, we should definitely do that: https://github.com/zooniverse/front-end-monorepo/issues/636

**Comment by beckyrother** [comment]
@rogerhutchings yes definitely happy to do that. Is there a link? 
Also, just saw the 'Why am I seeing this?' color conversation; please make that #fff. I'd hoped to keep the banner at a percentage opacity since it's on top of the image and I want to avoid confusion or missing elements if something gets covered up. 
I came across this plugin that might be useful: http://khan.github.io/tota11y/

**Comment by srallen** [comment]
Hm, the site I was using to calculate the ratio is saying the 0.8 opacity is worse.... I don't really understand that. FWIW, I tried to run the dev tools lighthouse audit and it didn't log in me when it executed the javascript for the audit, so it missed the banners. I'm wondering if we could setup a storybook for these? Could be done a separate PR since I don't think the classifier has storybook setup at all yet.

**Comment by rogerhutchings** [comment]
@srallen 

1. Grommet's animations were unsuitable, as you can't specify visibility independently of the animation, so on first view you get the Box animating out. I've replaced `react-pose` with a CSS animation.
1. I started down your suggested refactor path, and it just made everything more complicated; local state is required to stop weird text transition effects, and that means visibility needs to be handled separately from that.  If the hardest part to grok about the current setup is that the animation controls visibility, I can add a comment in the README, and we should punt a refactor to when we have more time.
1. _Presumably_ the hardest combination would be white text, on semi-transparent teal, on white? 
1. You could always force the banner to appear by modifying the `show` prop and then running lighthouse.
1. Storybook issue: https://github.com/zooniverse/front-end-monorepo/issues/639

@beckyrother No problem, I've set the text colour is white. On the accessibility review, I meant is there a checklist of things that we can check before the designs go to dev, so we can catch issues like contrast on transparency before they get to this point?

**Comment by srallen** [comment]
Thanks @rogerhutchings. Re:

1. That's a shame. They support multiple animation effects and you can set delays, but it sounds like it didn't work in this case? Thanks for trying it out. React-pose looks cool, but I also want to keep dependencies at a minimum if we can too.
2. Ok, I understand. Noting in a readme is an ok with me for interim and we can revisit for refactor at another time.
3. Yeah, a light colored subject would potentially make it hard to read. I don't think we have any that'll be pure white, but maybe a light grey scale like P4: Terrains: https://www.zooniverse.org/projects/mschwamb/planet-four-terrains/classify
I find the 0.8 opacity to be easier on the eyes, but we should get feedback from volunteers on this one.
4. The lighthouse audit re-executes the javascript so setting it in the react dev tools didn't work for me. I might be doing something wrong here.
5. Thanks!

I'll do a final review and this should be good to go.

**Comment by rogerhutchings** [comment]
Yeah, it was annoying; the way I did it was by setting a slide animation when `show` was true, and changing that to `fadeOut` on false; but because it's false to start off with, you get the animation happening on load, which sucked. I think those animations will work for other stuff, but show/hide isn't one of them.

On lighthouse testing, I meant if you actually change the code to force it to show, then run lighthouse, it should(? assuming the tests wait for everything to load first) cover it.

**Comment by srallen** [comment]
Oh, yeah that would work with lighthouse testing! 

I'm going to open an issue to enhance the explanation tooltip to include an option to select another workflow or perhaps a button to open the project/other workflow suggestion modal. We can build that out when we build out the rest of the workflow selection.

**Comment by beckyrother** [comment]
One thing I just noticed on the comps (it's been a while since I worked on this): The white text has a shadow, which will help if there's a very light subject:
image
The shadow would be `text-shadow: 0px 2px 2px #00000020`

**Comment by beckyrother** [comment]
BBC have a little accessibility checklist if that's what you're looking for? https://bbc-news.github.io/accessibility-news-and-you/accessibility-news-and-designers

**Comment by srallen** [comment]
@beckyrother I think @rogerhutchings means a more automated way like if there's a plugin or setting we can run in invision's inspect feature?

**Comment by beckyrother** [comment]
Got it. InVision doesn't offer any kind of plugins, unfortunately, but I have a Sketch plugin called [Stark](https://www.getstark.co/) that I use to check for color contrast and colorblindness. 

There's also [this chrome plugin ](http://khan.github.io/tota11y/) that looks neat, but it's for after the dev phase.

**Pull request merged**